File: tic-tac-toe/Python/tic-tac-toe.py
table = {
    "1": -1,
    "2": -1,
    "3": -1,
    "4": -1,
    "5": -1,
    "6": -1,
    "7": -1,
    "8": -1,
    "9": -1,
}

empty_table = table.copy()

player = False


def get_player_num():
    return 1 if player else 0


def print_table():
    def fmt(x, i):
        if x == 0:
            return "O"
        elif x == 1:
            return "X"
        return str(i)

    print(
        "".join(
            f"|{fmt(table[e], i)}|" + ("\n" if i % 3 == 0 else "")
            for i, e in enumerate(table, start=1)
        )
    )


def check_for_win():
    pnum = get_player_num()

    def check_vertical(offset):
        return (
            table[str(offset)] == pnum
            and table[str(offset + 3)] == pnum
            and table[str(offset + 6)] == pnum
        )

    def check_horizontal(offset):
        return (
            table[str(offset)] == pnum
            and table[str(offset + 1)] == pnum
            and table[str(offset + 2)] == pnum
        )

    def check_diagonal():
        return (table["1"] == pnum and table["5"] == pnum and table["9"] == pnum) or (
            table["3"] == pnum and table["5"] == pnum and table["7"] == pnum
        )

    vertical = any(check_vertical(i) for i in range(1, 4))
    horizontal = any(check_horizontal(i) for i in range(1, 10, 3))
    diagonal = check_diagonal()

    return vertical or horizontal or diagonal


def check_for_tie():
    return all(table[e] != -1 for e in table)


loop = True

while loop:
    while 1:
        print_table()

        print(f"Turn: Player {player+1}.")
        x = input("Pos: ")

        try:
            f = table[x]
        except:
            print("Err: Invalid pos.")
            continue

        if f != -1:
            print("Err: This field is taken.")
            continue

        table[x] = get_player_num()

        if check_for_win():
            print_table()
            print(f"Player {player+1} wins!")
            break

        if check_for_tie():
            print_table()
            print(f"Tie!")
            break

        player = not player

    while 1:
        x = input("Play again? (1/0): ")

        if x == "1":
            table = empty_table.copy()
            player = False
            break

        elif x == "0":
            loop = False
            print("Goodbye!")
            break

File: tic-tac-toe/JavaScript/tic-tac-toe.js
// start this with your browser
let table = {
	"1": -1,
	"2": -1,
	"3": -1,
	"4": -1,
	"5": -1,
	"6": -1,
	"7": -1,
	"8": -1,
	"9": -1
}

let empty_table = {
	...table
}

let player = false

function str(x)
{
	return '' + x // any -> str
}

function getPlayerNum()
{
	return player ? 1 : 0
}

function printTable()
{
	function format(x, i)
	{
		if (x == 0)
		{
			return "O"
		}
		else if (x == 1)
		{
			return "X"
		}
		return str(i)
	}

	let buffer = ""

	for (let i = 1; i < 10; i++)
	{
		buffer = buffer + `|${format(table[str(i)], i)}|`

		if (i % 3 == 0)
		{
			console.log(buffer)
			buffer = ""
		}
	}
}

function checkForWin()
{
	let i
	let pnum = getPlayerNum()

	function checkVertical(offset)
	{
		return (
			table[str(offset)] == pnum &&
			table[str(offset + 3)] == pnum &&
			table[str(offset + 6)] == pnum
		)
	}

	function checkHorizontal(offset)
	{
		return (
			table[str(offset)] == pnum &&
			table[str(offset + 1)] == pnum &&
			table[str(offset + 2)] == pnum
		)
	}

	function checkDiagonal()
	{
		return (
			table["1"] == pnum && table["5"] == pnum && table["9"] == pnum
		) || (
			table["3"] == pnum && table["5"] == pnum && table["7"] == pnum
		)
	}

	let vertical, horizontal, diagonal

	for (i = 1; i < 5; i++)
	{
		if (checkVertical(i))
		{
			vertical = true
			break
		}
	}

	for (i = 1; i < 10; i = i + 3)
	{
		if (checkHorizontal(i))
		{
			horizontal = true
			break
		}
	}

	diagonal = checkDiagonal()

	return vertical || horizontal || diagonal
}

function checkForTie()
{
	for (var e in table)
	{
		if (table[e] == -1)
		{
			return false
		}
	}
	return true
}

let loop = true

while (loop)
{
	while (true)
	{
		printTable()

		console.log(`Turn: Player ${player+1}.`)
		let x = prompt("Pos: ")

		f = table[x]

		if (f == undefined)
		{
			console.log("Err: Invalid pos.")
			continue
		}

		if (f != -1)
		{
			console.log("Err: This field is taken.")
			continue
		}

		table[x] = getPlayerNum()

		if (checkForWin())
		{
			printTable()
			console.log(`Player ${player+1} wins!`)
			break
		}

		if (checkForTie())
		{
			printTable()
			console.log("Tie!")
			break
		}

		player = !player

	}

	while (true)
	{
		x = prompt("Play again? (1/0): ")

		if (x == "1")
		{
			table = {
				...empty_table
			}
			player = false
			break
		}
		else if (x == "0")
		{
			loop = false
			console.log("Goodbye!")
			break
		}
	}
}
